File: backend/app/services/alert_service.py
# ─── SENTINEL Alert Service ───────────────────────────────────────────────────
# All alert CRUD operations. Routers call these — never query DB directly.

import json
from datetime import datetime, timezone
from typing import Optional
from sqlalchemy.orm import Session
from sqlalchemy import func

from app.models.alert import Alert
from app.models.user import User
from app.schemas.alert import AlertResponse, ShapFactor
from app.services.risk_service import factors_from_json


def _to_response(alert: Alert, db: Session) -> AlertResponse:
    """Convert an ORM Alert to an AlertResponse schema."""
    user = db.query(User).filter(User.id == alert.user_id).first()
    shap_raw = factors_from_json(alert.shap_values)
    shap = [ShapFactor(**f) for f in shap_raw]
    return AlertResponse(
        id=alert.id,
        user_id=alert.user_id,
        user_name=user.name if user else "Unknown",
        event_id=alert.event_id,
        title=alert.title,
        description=alert.description,
        type=alert.type,
        severity=alert.severity,
        status=alert.status,
        risk_score=alert.risk_score,
        ip_address=alert.ip_address,
        location=alert.location,
        device=alert.device,
        shap_values=shap,
        recommendation=alert.recommendation,
        created_at=alert.created_at,
        resolved_at=alert.resolved_at,
    )


def get_alerts(
    db: Session,
    severity: Optional[str] = None,
    status:   Optional[str] = None,
    type_:    Optional[str] = None,
    user_id:  Optional[str] = None,
    page:     int = 1,
    limit:    int = 20,
) -> dict:
    """Paginated, filtered alert list."""
    q = db.query(Alert)
    if severity: q = q.filter(Alert.severity == severity)
    if status:   q = q.filter(Alert.status == status)
    if type_:    q = q.filter(Alert.type == type_)
    if user_id:  q = q.filter(Alert.user_id == user_id)

    total     = q.count()
    open_c    = db.query(func.count(Alert.id)).filter(Alert.status == "open").scalar()
    critical  = db.query(func.count(Alert.id)).filter(Alert.severity == "critical").scalar()
    alerts    = q.order_by(Alert.created_at.desc()).offset((page - 1) * limit).limit(limit).all()

    return {
        "data":     [_to_response(a, db) for a in alerts],
        "total":    total,
        "page":     page,
        "pages":    max(1, -(-total // limit)),
        "open":     open_c or 0,
        "critical": critical or 0,
    }


def create_alert(
    db:         Session,
    user_id:    str,
    event_id:   Optional[str],
    title:      str,
    description: str,
    type_:      str,
    severity:   str,
    risk_score: int,
    ip_address: Optional[str],
    location:   Optional[str],
    device:     Optional[str],
    shap_factors: list[dict],
    recommendation: str,
) -> Alert:
    """Create and persist a new alert. Updates user alert counters."""
    alert = Alert(
        user_id=user_id, event_id=event_id, title=title,
        description=description, type=type_, severity=severity,
        status="open", risk_score=risk_score,
        ip_address=ip_address, location=location, device=device,
        shap_values=json.dumps(shap_factors),
        recommendation=recommendation,
    )
    db.add(alert)
    # Update user alert counters
    user = db.query(User).filter(User.id == user_id).first()
    if user:
        user.total_alerts += 1
        user.open_alerts  += 1
        user.risk_score    = risk_score
        user.risk_level    = severity
    db.commit()
    db.refresh(alert)
    return alert


def resolve_alert(db: Session, alert_id: str, resolver_id: str) -> Optional[Alert]:
    """Mark an alert as resolved."""
    alert = db.query(Alert).filter(Alert.id == alert_id).first()
    if not alert: return None
    alert.status      = "resolved"
    alert.resolved_at = datetime.now(timezone.utc)
    alert.resolved_by = resolver_id
    user = db.query(User).filter(User.id == alert.user_id).first()
    if user and user.open_alerts > 0:
        user.open_alerts -= 1
    db.commit()
    db.refresh(alert)
    return alert


def dismiss_alert(db: Session, alert_id: str) -> Optional[Alert]:
    """Mark an alert as dismissed."""
    alert = db.query(Alert).filter(Alert.id == alert_id).first()
    if not alert: return None
    alert.status = "dismissed"
    user = db.query(User).filter(User.id == alert.user_id).first()
    if user and user.open_alerts > 0:
        user.open_alerts -= 1
    db.commit()
    db.refresh(alert)
    return alert

File: frontend/src/services/alertService.ts
import type { Alert, RiskLevel, AlertStatus, AlertType } from '../types';
import { get, patch } from './api';

// ─── Types ────────────────────────────────────────────────────────────────────

export interface AlertFiltersDTO {
  severity?: RiskLevel;
  status?:   AlertStatus;
  type?:     AlertType;
  page?:     number;
  limit?:    number;
}

export interface PaginatedAlerts {
  data:  Alert[];
  total: number;
  page:  number;
  pages: number;
  open:  number;
  critical: number;
}

// ─── Mapper ───────────────────────────────────────────────────────────────────

export const mapAlert = (a: any): Alert => ({
  id: a.id,
  userId: a.user_id,
  userName: a.user_name || 'Monitored User',
  userEmail: a.user_email || '',
  eventId: a.event_id || '',
  title: a.title,
  description: a.description || '',
  type: a.type,
  severity: a.severity as RiskLevel,
  riskScore: a.risk_score,
  confidenceScore: a.confidence_score || 91,
  reasons: a.reasons || [a.description || 'Suspicious indicators detected.'],
  recommendedAction: a.recommendation || 'No action required.',
  status: a.status as AlertStatus,
  timestamp: a.created_at,
  resolvedAt: a.resolved_at || undefined,
  ipAddress: a.ip_address || 'N/A',
  device: a.device || 'N/A',
  location: a.location || 'N/A',
  country: a.location ? a.location.split(', ').pop() || 'IN' : 'IN',
  amount: a.amount,
  shapFactors: (a.shap_values || []).map((s: any) => ({
    feature: s.feature || 'unknown',
    label: s.factor || 'Factor',
    value: s.contribution || 0,
    displayValue: s.detail || (s.direction === 'positive' ? 'Increases risk' : 'Reduces risk')
  }))
});

// ─── Get all alerts (Admin) ───────────────────────────────────────────────────

export const getAlerts = async (
  filters: AlertFiltersDTO = {},
): Promise<PaginatedAlerts> => {
  const params: any = {};
  if (filters.severity && (filters.severity as any) !== 'all') params.severity = filters.severity;
  if (filters.status && (filters.status as any) !== 'all') params.status = filters.status;
  if (filters.type && (filters.type as any) !== 'all') params.type = filters.type;
  params.page = filters.page ?? 1;
  params.limit = filters.limit ?? 20;

  const res = await get<any>('/alerts/', params);
  return {
    data: (res.data || []).map(mapAlert),
    total: res.total,
    page: res.page,
    pages: res.pages,
    open: res.open || 0,
    critical: res.critical || 0,
  };
};

// ─── Get user alerts ──────────────────────────────────────────────────────────

export const getMyAlerts = async (
  filters: AlertFiltersDTO = {},
): Promise<PaginatedAlerts> => {
  const params: any = {};
  params.page = filters.page ?? 1;
  params.limit = filters.limit ?? 20;

  const res = await get<any>('/alerts/my', params);
  return {
    data: (res.data || []).map(mapAlert),
    total: res.total,
    page: res.page,
    pages: res.pages,
    open: res.open || 0,
    critical: res.critical || 0,
  };
};

// ─── Get single alert ─────────────────────────────────────────────────────────

export const getAlertById = async (id: string): Promise<Alert | null> => {
  try {
    const res = await get<any>(`/alerts/${id}`);
    return mapAlert(res);
  } catch {
    return null;
  }
};

// ─── Resolve alert ────────────────────────────────────────────────────────────

export const resolveAlert = async (id: string): Promise<void> => {
  await patch(`/alerts/${id}/resolve`);
};

// ─── Dismiss alert ────────────────────────────────────────────────────────────

export const dismissAlert = async (id: string): Promise<void> => {
  await patch(`/alerts/${id}/dismiss`);
};
